generic_config_updater/gu_common.py

```diff
@@ -174,3 +184,523 @@ def convert_sonic_yang_patch_to_config_db_patch(self, patch):
+        op_type = operation_type.name.lower()
+
+        operation = {OperationWrapper.OP_KEYWORD: op_type, OperationWrapper.PATH_KEYWORD: path}
+
+        if operation_type in [OperationType.ADD, OperationType.REPLACE]:
+            operation[OperationWrapper.VALUE_KEYWORD] = value
+
+        return operation
+
+class PathAddressing:
+    """
+    Path refers to the 'path' in JsonPatch operations: https://tools.ietf.org/html/rfc6902
+    The path corresponds to JsonPointer: https://tools.ietf.org/html/rfc6901
+
+    All xpath operations in this class are only relevent to ConfigDb and the conversion to YANG xpath.
+    It is not meant to support all the xpath functionalities, just the ones relevent to ConfigDb/YANG.
+    """
+    PATH_SEPARATOR = "/"
+    XPATH_SEPARATOR = "/"
+    XPATH_SEPARATOR_REPLACEMENT = "\n"
+    def get_path_tokens(self, path):
+        tokens = path.split(PathAddressing.PATH_SEPARATOR)[1:]
+
+        # src: https://tools.ietf.org/html/rfc6901
+        # > Because the characters '~' (%x7E) and '/' (%x2F) have special
+        # > meanings in JSON Pointer, '~' needs to be encoded as '~0' and '/'
+        # > needs to be encoded as '~1' when these characters appear in a
+        # > reference token.
+        mod_tokens = []
+        for token in tokens:
+            mod_token = str(token)
+            mod_token = mod_token.replace("~0", "~")
+            mod_token = mod_token.replace("~1", "/")
+            mod_tokens.append(mod_token)
+
+        return mod_tokens
+
+    def create_path(self, tokens):
+        if len(tokens) == 0:
+            return ""
+
+        # src: https://tools.ietf.org/html/rfc6901
+        # > Because the characters '~' (%x7E) and '/' (%x2F) have special
+        # > meanings in JSON Pointer, '~' needs to be encoded as '~0' and '/'
+        # > needs to be encoded as '~1' when these characters appear in a
+        # > reference token.
+        mod_tokens = []
+        for token in tokens:
+            mod_token = str(token)
+            mod_token = mod_token.replace("~", "~0")
+            mod_token = mod_token.replace("/", "~1")
+            mod_tokens.append(mod_token)
+
+        return f"{PathAddressing.PATH_SEPARATOR}{PathAddressing.PATH_SEPARATOR.join(mod_tokens)}"
+
+    def get_xpath_tokens(self, xpath):
+        """
+        Splits the given xpath into tokens by '/'.
+
+        Example:
+          xpath: /sonic-vlan:sonic-vlan/VLAN_MEMBER/VLAN_MEMBER_LIST[name='Vlan1000'][port='Ethernet8']/tagging_mode
+          tokens: sonic-vlan:sonic-vlan, VLAN_MEMBER, VLAN_MEMBER_LIST[name='Vlan1000'][port='Ethernet8'], tagging_mode
+        """
+        if xpath == "":
+            raise ValueError("xpath cannot be empty")
+
+        if xpath == "/":
+            return []
+
+        idx = 0
+        tokens = []
+        while idx < len(xpath):
+            end = self._get_xpath_token_end(idx+1, xpath)
+            token = xpath[idx+1:end]
+            tokens.append(token)
+            idx = end
```

Comment:
add UT with multiple token, but last element is empty i.e. /multi/tokens/with/empty/last/elem/